autoload/ale/completion.vim

```diff
@@ -262,6 +262,24 @@ function! ale#completion#Show(result) abort
         \   {-> ale#util#FeedKeys("\<Plug>(ale_show_completion_menu)")}
         \)
     endif
+
+    if l:source is# 'ale-callback'
+        call ale#completion#CallbackComplete()
+    endif
+endfunction
+
+function! ale#completion#CallbackComplete() abort
+    call b:asyncomplete_callback(b:ale_completion_result)
+endfunction
+
+function! ale#completion#GetAllTriggers() abort
+    return s:trigger_character_map
+endfunction
+
+function! ale#completion#RequestCallbackCompletions(callback) abort
+    let b:asyncomplete_callback = a:callback
+
+    call ale#completion#GetCompletions('ale-callback')
```

Comment:
I think instead of adding a wrapper around the `GetCompletions` function, add `...` to `GetCompletions` for varargs, and you can do the following.

```vim
let l:options = get(a:000, 0, {})

if len(a:000) > 1
    throw 'Too many arguments!'
endif

let l:Callback = get(l:options, 'callback', v:null)

if l:Callback isnot v:null
    let b:CompleteCallback = l:Callback
endif
```

You can then use `call ale#completion#GetCompletions('ale-callback', {'callback': CALLBACK_HERE})` instead below.

You should probably name funcref variables in PascalCase, as Vim typically doesn't let you name funcref variables as lowercase. I recommend using a more generic callback variable name here, for supporting other completion plugins later.